### PR TITLE
(sbt) Watch .proto files in protobuf folder to trigger PB.recompile

### LINKED ISSUE #148: Hook  .proto files to sbt's resource watch

### PR DESCRIPTION
**Pull request by ignasi35**
Fixes #148

This PR is heavily inspired by https://github.com/lagom/sbt-lagom-descriptor-generator/blob/master/lagom-descriptor-generator-sbt-plugin/src/main/scala/com/lightbend/lagom/spec/sbt/LagomOpenApiPlugin.scala#L22-L56 . There are several `sbt` details that I'm not familiar with and I'd rather have a second pair of eyes review this.

### CODE REVIEW

File: sbt-plugin/src/main/scala/akka/grpc/sbt/AkkaGrpcPlugin.scala [hunk 1]
```diff
@@ -38,13 +38,42 @@ object AkkaGrpcPlugin extends AutoPlugin {
       akkaGrpcCodeGenerators := GeneratorAndSettings(ScalaServerCodeGenerator, akkaGrpcCodeGeneratorSettings.value) :: Nil,
       akkaGrpcModelGenerators := Seq[Target]((JvmGenerator("scala", ScalaPbCodeGenerator), akkaGrpcCodeGeneratorSettings.value) -> sourceManaged.value),
 
+      PB.protoSources in Compile += target.value / "protobuf",
+
+      sourceDirectory in PB.recompile := sourceDirectory.value / "protobuf",
+      resourceDirectory in PB.recompile := sourceDirectory.value / "protobuf",
```
Comment (ignasi35) [inline sbt-plugin/src/main/scala/akka/grpc/sbt/AkkaGrpcPlugin.scala]: `PB.protoSources in Compile += target.value / "protobuf"` added here so users don't have to add it on their build.sbt.

`sourceDirectory.value / "protobuf"` hardcoded (or we can switch to `proto`) to prefer convention over config.

File: sbt-plugin/src/main/scala/akka/grpc/sbt/AkkaGrpcPlugin.scala [hunk 2]
```diff
@@ -38,13 +38,42 @@ object AkkaGrpcPlugin extends AutoPlugin {
...
+
+      target in PB.recompile :=
+        crossTarget.value / "protobuf" / Defaults.nameForSrc(configuration.value.name),
```
Comment (ignasi35) [inline sbt-plugin/src/main/scala/akka/grpc/sbt/AkkaGrpcPlugin.scala]: IIUC, this goes hand in hand with line `41`

File: sbt-plugin/src/main/scala/akka/grpc/sbt/AkkaGrpcPlugin.scala [hunk 3]
```diff
@@ -38,13 +38,42 @@ object AkkaGrpcPlugin extends AutoPlugin {
...
+      // TODO: review managedSources
+      managedSourceDirectories += (target in PB.recompile).value,
+      unmanagedResourceDirectories += (resourceDirectory in PB.recompile).value,
+      watchSources in Defaults.ConfigGlobal ++= (sources in PB.recompile).value,
```
Comment (ignasi35) [inline sbt-plugin/src/main/scala/akka/grpc/sbt/AkkaGrpcPlugin.scala]: This is the key setting.

**Comment by ignasi35** [comment]
Needs #145 or we copy `.travis.yml` here.

**Comment by ignasi35** [comment]
#145 was merged already. This is now rebased on top.

File: sbt-plugin/src/main/scala/akka/grpc/sbt/AkkaGrpcPlugin.scala [hunk 4]
```diff
@@ -38,13 +38,42 @@ object AkkaGrpcPlugin extends AutoPlugin {
       akkaGrpcCodeGenerators := GeneratorAndSettings(ScalaServerCodeGenerator, akkaGrpcCodeGeneratorSettings.value) :: Nil,
       akkaGrpcModelGenerators := Seq[Target]((JvmGenerator("scala", ScalaPbCodeGenerator), akkaGrpcCodeGeneratorSettings.value) -> sourceManaged.value),
 
+      PB.protoSources in Compile += target.value / "protobuf",
```
Comment (raboof) [inline sbt-plugin/src/main/scala/akka/grpc/sbt/AkkaGrpcPlugin.scala]: I'll have to take a bit of time to digest and play with this monday - haven't worked that closely with those plugins yet. Which task puts files into `target/protobuf`?

Comment (ignasi35) [reply sbt-plugin/src/main/scala/akka/grpc/sbt/AkkaGrpcPlugin.scala]: You are right, I think this is a copy/paste left-over we can get rid of.

Comment (ignasi35) [reply sbt-plugin/src/main/scala/akka/grpc/sbt/AkkaGrpcPlugin.scala]: To answer your question: no task puts files in `target/protobuf`.

**Comment by raboof** [comment]
I can confirm this works, but I'm not sure I understand all of the PR. This seems to basically-work by boiling it down to `watchSources in Defaults.ConfigGlobal ++= PB.protoSources.value,` - what are we missing when we do it like that? Something around caching it looks like?

**Comment by ignasi35** [comment]
> I can confirm this works, but I'm not sure I understand all of the PR. This seems to basically-work by boiling it down to watchSources in Defaults.ConfigGlobal ++= PB.protoSources.value, - what are we missing when we do it like that? Something around caching it looks like?

That works for regular sbt `~task`'s but won't work for [Lagom's devmode reloader](https://github.com/lagom/lagom/blob/55c5d1a92edf89c8ec2bd39b959e889c919627ee/dev/maven-plugin/src/main/scala/com/lightbend/lagom/maven/ServiceManager.scala#L105-L108) which watches the source folders using a different impl.

File: sbt-plugin/src/main/scala/akka/grpc/sbt/AkkaGrpcPlugin.scala (hunk 3 above)
Comment (ignasi35) [reply sbt-plugin/src/main/scala/akka/grpc/sbt/AkkaGrpcPlugin.scala]: Correction: these three lines are key (first two for dev-mode in play/lagom) and last for regular `~task`'s in `sbt`.

**Pull request merged**